[28기 최규흠] 레이아웃 구현, 유효성검사, 댓글기능 구현

EDGEPECT opened the pull request:
- Login 페이지 레이아웃 완성
- 유효성 검사 로직 작성

## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [O ] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
We스타그램 레이아웃, 로그인 유효성 검사, 댓글 기능구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1.https, CSS이용한 페이지 레이아웃 작성
2. 로그인 입력 데이터 유효성 검사 
계정 @포함 1글자 이상, 비밀번호 6자 이상 입력되지 않으면 버튼 disabled 되게 구현
opacity, disabled 속성 이용
3. 댓글 기능 구현,  게시버튼 클릭시 1글자 이상 입력 데이터 유효성 검사 이후 ul태그 아래에 li태그 새로 생성하여 innerhtml로 내용 추가, 

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
로그인 유효성검사는 어찌저찌 했는데 메인페이지 댓글기능 구현이 어려웠습니다.
세션에서 배운 로직을 토대로 접근하여 구현할 수 있었습니다.

<br />

## :: 기타 질문 및 특이 사항

pepekim commented:
깃허브 초기세팅완료

pepekim commented:
login.js 에서 input_id, input_pass로 나누지말고 querySelectorAll(input)으로 전부 먹이면 될것 같은데 작동하지가 않네요

pepekim commented:
main.js, login.js 코드 함수화 및 중복 제거

claire-euni commented on students/14th/kimtaehyun/js/login.js:
```diff
@@ -0,0 +1,21 @@
+    button.style.opacity = 1;
+}
+
+input[0].value.length < 0 ? input[0].addEventListener('keyup', () => {
+    button.style.opacity = 0.35;
+}) : input[0].addEventListener('keydown', () => {
+    make_blue();
+});
+
+input[1].value.length < 0 ? input[1].addEventListener('keyup', () => {
+    button.style.opacity = 0.35;
+}) : input[1].addEventListener('keydown', () => {
+    make_blue();
+});
```
- input Id & password 부분을 한번에 이벤트 적용하는것을 질문하셨는데 이런 경우는 input태그를 모두 감싼 상위 태그 input_container한테 이벤트를 적용해주면 이벤트 캡쳐링이 일어나서 한번에 자식에게도 이벤트가 적용됩니다. 이벤트 캡쳐링과 이벤트 버블링 한번 검색해보시는것 추천합니다.
- 현재 구현하신 기능에서는 input value가 썼을때는 파란색으로 버튼이 활성화됐다가 input값을 다시 지워서 value가 없는 경우에도 여전히 파란색을 유지합니다. 원하시는 방향인가요?

아래의 코드를 참고해보세요 🤗

```
const inputBox = document.querySelector(".input_container");

function activateBtn() {
  (input[0].value.length && input[1].value.length) != 0
    ? (button.style.opacity = 1)
    : (button.style.opacity = 0.35);
}

inputBox.addEventListener("keyup", activateBtn);
```

claire-euni commented on students/14th/kimtaehyun/style/main.css:
```diff
@@ -0,0 +1,512 @@
+    margin-right: 200px;
+}
+
+.navi_icon {
+    width: 30px;
+    height: 30px;
+    margin-right: 30px;
+    margin-left: 0px;
+}
+
+.heart_icon {
+    width: 30px;
+    height: 30px;
+    margin-right: 30px;
+}
+
+.man_icon {
+    width: 30px;
+    height: 30px;
+    margin-right: 30px;
+}
```
동일한 속성을 가진 셀렉터가 반복되고 있습니다. 한번에 묶어서 공통으로 적용되게 해주거나, 아예 세 개의 태그가 공통된 하나의 클래스명 navi_icon을 갖게 해주세요~
```css
.navi_icon, .heart_icon, .man_icon {
    width: 30px;
    height: 30px;
    margin-right: 30px;
    margin-left: 0px;          // <-필요하지 않은 속성같습니다.
}


```

claire-euni commented on students/14th/kimtaehyun/style/main.css:
```diff
@@ -0,0 +1,512 @@
+    margin-right: 30px;
+}
+
+main {
+    display: flex;
+    justify-content: space-between;
+}
+
+.feeds {
+    margin-left: 300px;
+}
+
+article {
+    display: flex;
+    flex-direction: column;
+    margin: 20px auto;
+    background-color: #fff;
+    border: 1px solid #ddd;
+}
+
+.profile_container {
+    display: flex;
+    justify-content: space-between;
+}
```
main과 .profile_container도 공통 속성이므로 한번에 적어줄 수 있겠죠?

claire-euni commented on students/14th/kimtaehyun/style/main.css:
```diff
@@ -0,0 +1,512 @@
+    cursor: pointer;
+}
+
+#feedImage {
+    width: 100%;
+    height: 620px;
+}
+
+.article_bottom {
+    display: flex;
+    flex-direction: column;
+    width: 100%;
+    background-color: #fff;
+}
```
모두 없어도 나타나는게 달라지지 않네요. 중복되는 속성이거나 필요없는 속성같습니다.  확인해주세요 :)

claire-euni commented on students/14th/kimtaehyun/style/main.css:
```diff
@@ -0,0 +1,512 @@
+    justify-content: flex-start;
+}
+
+.article_left_icons img {
+    display: flex;
+    justify-content: flex-start;
+    margin: auto 10px;
+    width: 30px;
+    height: 30px;
+    cursor: pointer;
+}
+
+.right_icon {
+    width: 30px;
+    height: 30px;
+    margin: auto 10px;
+    cursor: pointer;
+}
```
이 부분도 중복!

claire-euni commented on students/14th/kimtaehyun/style/main.css:
```diff
@@ -0,0 +1,512 @@
+.etc {
+    color: lightgray;
+    font-weight: bold;
+}
```
전체적으로 중복되거나 필요치않은 속성이 있는지 확인해주세요. 개발자도구에서 체크박스로 확인해보면 쉽답니다!
css에서 변수 설정하는 방법도 참고해보세요 (https://developer.mozilla.org/ko/docs/Web/CSS/var)

claire-euni commented on students/14th/kimtaehyun/js/main.js:
```diff
@@ -0,0 +1,38 @@
+const enterEvent = document.querySelector('#input_comment');
+const clickEvent = document.querySelector('post_comment_btn');
+const comment_container = document.querySelector('.comment_container');
+let now = new Date();
```
Date()도 이용하셨네요 와우👏👏

claire-euni commented on students/14th/kimtaehyun/js/main.js:
```diff
@@ -0,0 +1,38 @@
+const enterEvent = document.querySelector('#input_comment');
+const clickEvent = document.querySelector('post_comment_btn');
```
변수명 네이밍 다시 고민해보세요! input_comment와 enterEvent는 상관성이 없어보여요~

claire-euni commented on students/14th/kimtaehyun/js/main.js:
```diff
@@ -0,0 +1,38 @@
+        newcomments.innerHTML = `<p>b2ng_9<span>${enterEvent.value}</span></p>`;
+        comment_container.append(newcomments);
+        enterEvent.value = '';
+    } else if(newcomments.innerHTML !== '') {
+        newcomments.innerHTML = `<p>b2ng_9<span>${enterEvent.value}</span></p>`;
+        comment_container.append(p);
+        enterEvent.value = '';
+    }
```
- else를 사용해보세요
- ` comment_container.append(p);` (p)가 무엇을 의미하나요? 
- 현재 comment 변수는 선언만 되고 사용되지 않고 있습니다. 필요한 변수인가요~?

claire-euni commented on students/14th/kimtaehyun/js/main.js:
```diff
@@ -0,0 +1,38 @@
+function post_comment() {
+    let newcomments = document.createElement('newcomments');
+    const comment = document.querySelector('.article_comments_1');
+    if (newcomments.innerHTML === '') {
+        newcomments.innerHTML = `<p>b2ng_9<span>${enterEvent.value}</span></p>`;
+        comment_container.append(newcomments);
```
innerHTML, append(), createElement() 사용 잘하셨습니다 👏

claire-euni commented on students/14th/kimtaehyun/js/main.js:
```diff
@@ -0,0 +1,38 @@
+    }
+}
+
+function time_ticking() {
+    const time = document.querySelector('time');
+    let writeDay = new Date();
+    let passedTime = Math.floor(parseInt(now - writeDay) / 1000);
+    time.innerHTML = passedTime + "초 전";
+    now = new Date();
+}
```
오 이런 디테일한 기능까지 구현하시고 👍👍

claire-euni commented on students/14th/kimtaehyun/main.html:
```diff
@@ -0,0 +1,126 @@
+                    </time>
+                    <div class="input_box">
+                      <input id="input_comment" placeholder="댓글 달기...">
+                      <post_comment_btn>게시</post_comment_btn>
```
이게 무슨 태그인가요.......?

pepekim replied:
수정했습니다

claire-euni commented on students/14th/kimtaehyun/main.html:
```diff
@@ -0,0 +1,126 @@
+                    <div class="comment_container"> 
+                      <div class="article_comments_1">
+                        <p>canon_mj<span>위워크에서 진행한 베이킹 클래스...</span><more>더 보기</more></p>
+                        <comment_heart>
+                          <img src="https://s3.ap-northeast-2.amazonaws.com/cdn.wecode.co.kr/bearu/heart.png">
+                        </comment_heart>
```
태그가 comment_heart 인가요.....?

claire-euni reviewed (commented):
태현님 첫 PR 잘해주셨습니다! 리뷰 반영해서 리팩토링해주세요 😄🤟

pepekim commented on students/14th/kimtaehyun/login.html:
```diff
@@ -0,0 +1,26 @@
+<head>
+	<meta charset="utf-8">
+	<meta name="viewport" content="width=device-width">
+	<title>repl.it</title>
```
수정했습니다

pepekim commented on students/14th/kimtaehyun/login.html:
```diff
@@ -0,0 +1,26 @@
+	<link href="style/login.css" rel="stylesheet" type="text/css" />
+</head>
+<body>
+	<div class="container">
+    <div class="up_container">
```
수정했습니다

pepekim commented on students/14th/kimtaehyun/js/login.js:
```diff
@@ -0,0 +1,18 @@
+function make_blue() {
+    let button = document.querySelector('button');
+    button.style.opacity = 1;
+}
```
다 지웠을때 다시 버튼 색깔이 옅어지는게 쉽지 않네요
이것저것 테스트 해보고 있는데 한번 봐주시겠어요?

console.log(input[0].value); // 빈칸
console.log(input[0].value.length); // 0이 출력됨

if(input[0].value.length === 0) { // 
    deactivate_Btn(); // <= 버튼을 옅게 만드는 함수
} else if (input[0].value.length !== 0){
    activate_Btn(); // <= 버튼을 진하게 만드는 함수
}

제 생각으로는 아무것도 입력을 안하면 console창에서 보이듯이 input[0].value.length === 0 이니까 deactivate_Btn 함수가 작동하면서 옅어지고 입력을 하면은 input[0].value.length가 0이 아니니까 activate_Btn 함수가 작동하면서 다시 진해져서 작동하는게 맞는것 같은데 왜 안되는지 모르겠어요

input[0].value.length < 0 ? button.style.opacity = 0.35 : input[0].addEventListener('keyup', () => {
    activate_Btn();
});
이거는 잘 작동하는데요
input[0].value.length > 0 ? input[0].addEventListener('keyup', () => {
    activate_Btn();
}) : button.style.opacity = 0.35;
이거는 왜 작동하지 않는지 궁금합니다 부등호랑 안에 관계만 그대로 바꿨는데 왜 저건 되지 이건 안되죠??

제가 생각했을때 최종적으로 이렇게 식을 짜면 될것 같은데요

input.value.length가 0이 되는 조건 ? 버튼을 연하게 하는 식 : 버튼을 찐하게 하는 식
이런식으로 하면 될것 같은데요 디테일한 부분을 잘 모르겠습니다.

pepekim commented on students/14th/kimtaehyun/js/login.js:
```diff
@@ -0,0 +1,18 @@
+let input_id = document.querySelector('.input_id');
+let input_pass = document.querySelector('.input_pass');
+
+// input도 id, pass 나누지말고 하나로 주고싶은데 왜 안되지?? querySelectorAll 하면될거 같은데..
```
알겠습니다

pepekim commented:
반복되는 코드를 반복문으로 돌리는 방법을 잘 모르겠어요ㅠ

EDGEPECT committed:
[Add] Login 페이지 레이아웃 & 유효성 검사 작성중
- Login 페이지 레이아웃 완성
- 유효성 검사 로직 완성//

EDGEPECT committed:
로그인 페이지 , 메인페이지 레이아웃 구현중

EDGEPECT committed:
<build>: <레이아웃 및 필수사항 구현>

로그인 페이지, 메인 페이지 레이아웃 구현,
로그인 페이지 아이디에 @ 포함, 1글자 이상일 것 + 비밀번호 6자 이상일것 조건으로 login버튼 활성/비활성화
메인 페이지 댓글 입력칸 1글자 이상일시 활성화되고 게시 버튼 클릭시 댓글 추가.
